File: Leetcode/DP/85. Maximal Rectangle.cpp
class Solution {
public:
    int maximalRectangle(vector<vector<char>>& matrix) {
        int n = matrix.size(), m = matrix[0].size();
        int dp[205][205][205][2] = {0}, ans = 0;

        for(int i = 0; i < n; i++) {
            for(int j = 0; j < m; j++) {
                if(matrix[i][j] == '0') continue;

                for(int k = j; k >= 0; k--) {
                    if(matrix[i][k] == '0') break;

                    dp[i][j][k][0] = (i-1 < 0) ? j-k+1 : dp[i-1][j][k][0] + j-k+1;
                    ans = max(ans, dp[i][j][k][0]);
                }

                for(int k = i; k >= 0; k--) {
                    if(matrix[k][j] == '0') break;

                    dp[i][j][k][1] = (j-1 < 0) ? i-k+1 : dp[i][j-1][k][1] + i-k+1;
                    ans = max(ans, dp[i][j][k][1]);
                }
            }
        }

        return ans;
    }
};

File: Leetcode/DP/85. Maximal Rectangle.py
class Solution:
    def maximalRectangle(self, matrix: List[List[str]]) -> int:
        m, n = len(matrix), len(matrix[0])
        cum = [[0 for _ in range(n)] for _ in range(m)]

        for i in range(m):
            for j in range(n-1, -1, -1):
                if j == n-1:
                    cum[i][j] = int(matrix[i][j])
                else:
                    cum[i][j] = 0 if matrix[i][j] == '0' else int(matrix[i][j]) + cum[i][j+1]
        
        ans = 0
        for i in range(m):
            for j in range(n):
                width = cum[i][j]
                for k in range(i, m):
                    width = min(width, cum[k][j])
                    ans = max(ans, (k-i+1)*width)
        
        return ans